program/src/processor/initialize_config.rs

```diff
@@ -65,24 +66,56 @@ pub fn process_initialize_config(
     // 2. vault (token account)
     // - must be initialized
     // - have the vault signer (PDA) as owner
+    // - no close authority or delegate
+    // - no other extension apart from `ImmutableOwner` and `TransferHookAccount`
     // - have the correct mint
     // - amount equal to 0
 
     let vault_data = ctx.accounts.vault.try_borrow_data()?;
     // unpack checks if the token is initialized
     let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
 
-    let (vault_signer, bump) = Pubkey::find_program_address(
-        &["token-owner".as_bytes(), ctx.accounts.config.key.as_ref()],
-        program_id,
-    );
+    let (vault_signer, signer_bump) = find_vault_pda(ctx.accounts.config.key, program_id);
 
     require!(
         vault.base.owner == vault_signer,
         StakeError::InvalidTokenOwner,
         "vault"
     );
 
+    require!(
+        vault.base.close_authority == PodCOption::none(),
+        StakeError::CloseAuthorityNotNone,
+        "vault"
+    );
+
+    require!(
+        vault.base.delegate == PodCOption::none(),
+        StakeError::DelegateNotNone,
+        "vault"
+    );
+
+    let mut required_extensions = Vec::from(VALID_VAULT_TOKEN_EXTENSIONS);
+
+    vault
+        .get_extension_types()?
+        .iter()
+        .try_for_each(|extension_type| {
+            if !VALID_VAULT_TOKEN_EXTENSIONS.contains(extension_type) {
+                msg!("Invalid token extension: {:?}", extension_type);
+                return Err(StakeError::InvalidTokenAccountExtension);
+            }
+            required_extensions.retain(|e| e != extension_type);
```

Comment:
Maybe I'm misunderstanding the idea here, but if the token account has both extensions, then I think this will actually remove all the `required_extensions` and fail the check later, which isn't the intention.

For example, if `get_extension_types()` returns `[TransferHookAccount, ImmutableOwner]`, then the first iteration will remove `ImmutableOwner`, and the second iteration will remove `TransferHookAccount`.

I think we can get rid of the `required_extensions` check, but keep the `VALID_VAULT_TOKEN_EXTENSIONS` check.

Reply:
Hehe looks like I did totally misunderstand, sorry about that.

Either way, let's keep it simple and only allow those two extensions, immutable ownership shouldn't be required.